### ISSUE TITLE
The performance of nginx drops a lot when using OpenSSL 3.0

### ISSUE THREAD

**Issue by ShuaiYuan21** [openssl/openssl#21833]
### Some parameters:
Cipher suite: TLS_AES_256_GCM_SHA384
TLS version: TLSv1.3
Kx=EC
Au=EC
ECDH_curve=prime256v1

### CPS test result:
| OpenSSL 1.1.1 | OpenSSL 3.0.9 | drop  |
|---------------|-------------|--------|
| 4295.38       | 3132.76     | 27.07% |

### Flame Graph:
nginx with OpenSSL 1.1.1:
![sw_ec_ossl1_sw](https://github.com/openssl/openssl/assets/84753943/ed417919-befc-4ce1-8e9d-fd13127c8f47)

nginx with OpenSSL 3.0.9
![sw_ec_ossl3_sw](https://github.com/openssl/openssl/assets/84753943/2e3ed506-7005-47e7-81af-0e1c6e3f1a05)

### Root Cause
**1. EVP_PKEY_derive_set_peer_ex**
<img width="374" alt="image" src="https://github.com/openssl/openssl/assets/84753943/dfa6c347-3978-4622-a78e-aed73ed28f37">

I don't know if this part of the operation is necessary. 
But it can be seen that it takes up a lot of CPU resources, so in order to eliminate its impact on performance, we can disable it in the source code.

```Bash
diff --git a/crypto/evp/exchange.c b/crypto/evp/exchange.c
index d7a4ad142a..58eea73b80 100644
--- a/crypto/evp/exchange.c
+++ b/crypto/evp/exchange.c
@@ -499,7 +499,7 @@ int EVP_PKEY_derive_set_peer_ex(EVP_PKEY_CTX *ctx, EVP_PKEY *peer,

 int EVP_PKEY_derive_set_peer(EVP_PKEY_CTX *ctx, EVP_PKEY *peer)
 {
-    return EVP_PKEY_derive_set_peer_ex(ctx, peer, 1);
+    return EVP_PKEY_derive_set_peer_ex(ctx, peer, 0);
 }
```
Performance comparison after EVP_PKEY_derive_set_peer_ex disabled 
| OpenSSL 1.1.1 | OpenSSL 3.0.9 | drop  |
|---------------|-------------|--------|
| 4295.38       | 3747.19     | 12.76% |

The flame graph after EVP_PKEY_derive_set_peer_ex disabled 
![sw_ec_ossl3_sw_no_set_peer](https://github.com/openssl/openssl/assets/84753943/f62c07d9-3336-47dd-8ffa-44fa1ad04c5e)

Now, it looks good, and the performance improved a lot.

Below are based on the flame graph that disabled the EVP_PKEY_derive_set_peer_ex

**2. do_sigver_init**
This function takes up ~4% of CPU resources.
<img width="228" alt="image" src="https://github.com/openssl/openssl/assets/84753943/ab535cf1-065a-4363-a8a5-518bdc2bf517">

**3. EVP_xxx_fetch**
These functions take up ~6.3% of CPU resources. 
In OpenSSL, these functions are similar to engine_table_select, which consumes fewer CPU cycles.
<img width="496" alt="image" src="https://github.com/openssl/openssl/assets/84753943/6d1a795e-c397-4a0b-95ec-890a10069465">

Most of it is a **lock** operation.
<img width="892" alt="image" src="https://github.com/openssl/openssl/assets/84753943/c4d30bc4-c5ea-4b2e-93f9-9f6506342148">

**4. Some other code changes**

### Question
1. If EVP_PKEY_derive_set_peer_ex is not necessary, can we disable it when config the project?
2. Will the function do_sigver_init and EVP_xxx_fetch be optimized in the future release?

**Comment by t8m** [comment]
The problem is that without the validate_peer parameter set the public key of the peer is not validated which means it can be a bogus key not providing security against eavesdroppers or mitm.

**Comment by t8m** [comment]
However perhaps EVP_PKEY_public_check_quick() would be sufficient instead of the full check. @slontis ?

**Comment by davidben** [comment]
Looks like the "quick" check is checking that the point isn't infinity, checking coefficients are in range, and checking if the point is on a curve. Other than potentially the first one (depending on your invariants), all these checks ought to have been performed at the time you constructed the `EC_KEY`. I.e. OpenSSL should not allow you to construct an `EC_POINT` that is off the curve... I thought you all fixed that in 1.1.x or so? But it's also not that expensive to redo wastefully.

The "full" check additionally checks that point * order = infinity. This one _is_ expensive, as you're effectively doing twice as much ECDH as you need. It is also completely pointless if you're on a prime-order curve, where there are no small subgroups, and practically all curves used in this part of the code have prime order.

**Comment by davidben** [comment]
> practically all curves used in this part of the code have prime order

Though, since OpenSSL supports arbitrary curves as well as a some built-in curves with a cofactor, keep in mind that when you *aren't* on a prime-order curve, you may need to take some care here.

**Comment by slontis** [comment]
Section 5.6.2.2.2 of SP800-56Ar3 says you can do either the partial OR the full public key validation, so changing to the partial check should be sufficient.

It is also not helped currently by the code performing the point multiply twice..

**Comment by slontis** [comment]
Perhaps if we know that there is no way that an EC_KEY can exist without already having done most of the checks, then maybe we could remove some of the redundant checks (such as pointoncureve) and comment why they are removed (even if it was only for the partial case).

**Comment by t8m** [comment]
> Though, since OpenSSL supports arbitrary curves as well as a some built-in curves with a cofactor, keep in mind that when you _aren't_ on a prime-order curve, you may need to take some care here. (I don't know that part as well because BoringSSL just doesn't support `EC_GROUP` with cofactors.)

We could do the quick check for prime order curves, otherwise do the full check.

**Comment by wbl** [comment]
We've also run into the use of the expensive test instead of quick check at Akamai, in the course of our 3.0 migration. I don't however have a patch that does the right thing with non-NIST curves (although x25519 should bypass all of this).

**Comment by wbl** [comment]
For find_sig_alg optimization the problem is that the useable hash functions are not recorded in the EVP_PKEY, and so the calls are expensive. Optimizing by saying if you use a named curve you probably support the right hash is one possibility and is what I tried, but it would be better if the EVP_PKEY could keep that information around v. need an expensive provider call every time.

**Comment by testn** [comment]
any updates? It seems worrisome.

**Comment by vyasatavi** [comment]
Will this fix be back ported to 3.0.x for LTS releases?

**Comment by t8m** [comment]
> Will this fix be back ported to 3.0.x for LTS releases?

I am sorry but no, we do not plan to backport any exisiting performance fixes to the 3.0 branch. The fixes are too invasive to be backported.
